[core][RF] New ROOT/RRangeCast.hxx with RangeStaticCast and RangeDynCast and first use of it in RooFit

guitargeek opened the pull request:
In RooFit, iterating over collections of `RooAbsArg*` and applying `static_cast` and `dynamic_cast` to every item is ubiquitous.

This PR introduces a way to do the casing in a less verbose way in the `for`-line of the range-based loop with the `static_span_cast` and `dynamic_span_cast` functions. Thanks to @hageboeck for the suggestion.

The second commit in this PR shows examples on how they can be used in RooFit.

The reason for doing this PR now it that I want to open a `good-first-issue` in RooFit range-based loop migration for people who want to contribute to RooFit by doing some easy tasks. These span-casting functions are the missing ingredient before the migration can be started.

By the way, the names of the functions were inspired by `static_pointer_cast` and `dynamic_pointer_cast`:
https://en.cppreference.com/w/cpp/memory/shared_ptr/pointer_cast

pcanal commented:
How do these compare/contrast with the existing `TRangeDynCast`, `TRangeStaticCast` (and/or `TRangeDynCastIterator`, etc.). See `TCollection.h`?

guitargeek commented:
Thanks for pointing that out, I didn't know about these existing classes!

It appears `TRangeDynCast` and `TRangeStaticCast` can only be used with TCollections:
https://github.com/root-project/root/blob/master/core/cont/inc/TCollection.h#L419

The RooAbsCollection we want to use this collection casting for doesn't inherit from TCollection (see the inheritance graph):
https://root.cern/doc/v608/classRooAbsCollection.html

eguiraud reviewed (changes requested):
Hey, I left some comments, but the only ones that matter are those related to "iterator-concept compliance": I think `SpanCaster::begin` and `SpanCaster::end` should a) return an iterator type that satisfies at least the `LegacyInputIterator` concept and b) allow const as well as non-const iteration.

Otherwise users will have bad surprises and/or terrible compilation errors when they try to pass a `dynamic_span_cast`'d collection to a STL algorithm, for instance.

guitargeek commented on roofit/roofitcore/inc/SpanCast.h:
```diff
@@ -0,0 +1,111 @@
+// Author: Jonas Rembser, CERN, Jul 2021
```
I don't know what the rules are here, it seems that everybody does what he or she wants in new RooFit files. So far, I have not added some header to new files and @lmoneta didn't complain about out :)

hageboeck replied:
There's no "rules" as such, but all RooFit files that are not tests do have a header ...

hageboeck replied:
... excluding a few external contributions maybe.

guitargeek replied:
Okay fine I don't mind putting a header, but which one? The old RooFit header or the general ROOT header? I see many files with some special RooFit header: https://github.com/root-project/root/blob/master/roofit/roofitcore/inc/RooAbsBinning.h.

But I also see files that you added with the ROOT header: https://github.com/root-project/root/blob/master/roofit/roofitcore/inc/RooWorkspaceHandle.h.

Should I put the ROOT header then?

hageboeck replied:
> Okay fine I don't mind putting a header, but which one? The old RooFit header or the general ROOT header? I see many files with some special RooFit header: https://github.com/root-project/root/blob/master/roofit/roofitcore/inc/RooAbsBinning.h.

That should be most files, actually.

> 
> But I also see files that you added with the ROOT header: https://github.com/root-project/root/blob/master/roofit/roofitcore/inc/RooWorkspaceHandle.h.
That's my 2nd month. Maybe I hadn't decided by then ... 😅 

> Should I put the ROOT header then?
Given the above: No?

hageboeck replied:
> I just remembered that `TRangeDynCast` and `TRangeStaticCast` are already a thing in ROOT, meant for use with `TCollection`s. I looked them up and I think they are generic enough to be used with RooFit collections too (but they suffer from similar non-STL-compliance issues as the first version of this PR). Maybe generalize/improve those rather than adding a RooFit-specific version? So everybody benefits.

I actually pointed specifically to those when bringing this up, so 👍. And yes, they need some love. And they also should be documented much better, see #8421.
I think it's fair to adapt them properly for use in RooFit, because then they will be used more actively, and get the love they deserve.

guitargeek commented on roofit/roofitcore/inc/SpanCast.h:
```diff
@@ -0,0 +1,111 @@
+#ifndef roofit_roofitcore_SpanCast_h
+#define roofit_roofitcore_SpanCast_h
+
+namespace detail {
```
Okay, `ROOT::Detail` is fine!

guitargeek commented on roofit/roofitcore/inc/SpanCast.h:
```diff
@@ -0,0 +1,111 @@
+   Span_t const &_inputSpan;
+};
+
+/// Takes any span that can be used in range-based loops and applies
```
Okay, I changed "span" to "range" everywhere!

guitargeek commented on roofit/roofitcore/inc/SpanCast.h:
```diff
@@ -0,0 +1,111 @@
+
+namespace detail {
+
+template <bool isDynamic, typename T>
+struct TemplatedCaster {
+   template <typename U>
+   static T cast(U const &u);
+};
+
+template <typename T>
+struct TemplatedCaster<false, T> {
+   template <typename U>
+   static T cast(U const &u)
+   {
+      return static_cast<T>(u);
+   }
+};
+
+template <typename T>
+struct TemplatedCaster<true, T> {
+   template <typename U>
+   static T cast(U const &u)
+   {
+      return dynamic_cast<T>(u);
+   }
+};
```
I don't this this solves the problem of deciding on `dynamic_cast` vs `static_cast` at compile time, no? In C++ it would be so easy with `if constexpr` ... I added some more comments to explain why the boilerplate code is needed.

eguiraud replied:
> I don't this this solves the problem of deciding on dynamic_cast vs static_cast at compile time,

yes it should (although constexpr-if is best): https://godbolt.org/z/dj3zshjqW

If you anyway prefer structs instead of functions+tag types, you can reduce the boilerplate a bit by declaring only one specialization instead of 2:

```cpp
template <bool isDynamic = false, typename T>
struct TemplatedCaster {
   static T cast(U const &u)
   {
      return static_cast<T>(u);
   }
};

template <typename T>
struct TemplatedCaster<true, T> {
   template <typename U>
   static T cast(U const &u)
   {
      return dynamic_cast<T>(u);
   }
};
```

hageboeck replied:
My take in this:
https://godbolt.org/z/18ndvr35x

hageboeck commented on roofit/roofitcore/inc/RangeCast.h:
```diff
@@ -0,0 +1,111 @@
+
+namespace detail {
+
+template <bool isDynamic, typename T>
+struct TemplatedCaster {
+   template <typename U>
+   static T cast(U const &u);
+};
+
+template <typename T>
+struct TemplatedCaster<false, T> {
+   template <typename U>
+   static T cast(U const &u)
+   {
+      return static_cast<T>(u);
+   }
+};
+
+template <typename T>
+struct TemplatedCaster<true, T> {
+   template <typename U>
+   static T cast(U const &u)
+   {
+      return dynamic_cast<T>(u);
+   }
+};
```
I think this can be removed.

hageboeck commented on roofit/roofitcore/inc/RangeCast.h:
```diff
@@ -0,0 +1,111 @@
+      }
+      bool operator==(const CastingIterator &rhs) const { return _iter == rhs._iter; }
+      bool operator!=(const CastingIterator &rhs) const { return _iter != rhs._iter; }
+      T operator*() { return detail::TemplatedCaster<isDynamic, T>::cast(*_iter); }
```
I assume that the compiler will do equally well as for `if constexpr` if you use this:
```c++
if (isDynamic)
  return dynamic_cast<T>(*_iter);
else
  return static_cast<T>(*_iter);
```
You can do similarly for `operator->`.

hageboeck replied:
Further, think about
```c++
else {
  assert(dynamic_cast<T>(*_iter));
  return static_cast<T>(*_iter);
}
```
This might save you from making a few not-so-easy-to-find mistakes.

guitargeek replied:
Good idea!

eguiraud replied:
Note that `static_cast` is free in terms of runtime, `dynamic_cast` (used e.g. in the `assert` above) might add a little overhead when RTTI is required which users of `static_cast` might not expect.

(I don't know if these adapters will be used in hot loops a lot, so might be not relevant, feel free to ignore)

guitargeek replied:
These will be used in hot loops. But the RooFit philosophy so far was to ignore the performance penalty intoduced by `assert` because it is ignored in the release builds.

I see now that this is an unexpected behavior if `static_range_cast` is used by people other than the RooFit developers, so actually I think it's better to not do the assert here.

hageboeck replied:
Your decision. I'm just saying that those asserts helped trace down a ton of bugs, and it's really as easy as running the thing that somebody sends you in your debug build.

eguiraud replied:
sorry I take back my comment about performance, `assert`s are totally fine in performance-critical code!

hageboeck commented on roofit/roofitcore/src/RooAbsData.cxx:
```diff
@@ -2168,10 +2168,7 @@ void RooAbsData::optimizeReadingWithCaching(RooAbsArg& arg, const RooArgSet& cac
-      RooRealVar* rrv = dynamic_cast<RooRealVar*>(obs) ;
+    for(auto const* rrv : dynamic_range_cast<RooRealVar*>(*usedObs)) {
       if (rrv && !rrv->getBinning().isShareable()) {
    RooArgSet depObs ;
```
Maybe fix indentation in 2nd commit if you touch this loop.

guitargeek replied:
> This looks quite good already, but consider removing the helper class.
> 
> ```c++
> if (templateParameter)
> ```
> 
> won't have any runtime overhead with any decent compiler, and both `dynamic_cast` and `static_cast` are syntactically correct code paths, so no `if constexpr` required.

Thanks for the comments! Before I address the other comments: just one answer to this comment. No, it seems that `dynamic_cast<T*>` is not syntactically correct if T is not polymorphic, e.g. it has not virtual method.

You can for example see that this code doesn't compile:

```C++
struct A {
    void foo() {}
};

struct B : public A {
    void bar() {}
};

int main() {

    A* ptr = new B;
    dynamic_cast<B*>(ptr);
    delete ptr;
}
```
Here is the error:
```
collection_cast.cc: In function ‘int main()’:
collection_cast.cc:95:5: error: cannot ‘dynamic_cast’ ‘ptr’ (of type ‘struct A*’) to type ‘struct B*’ (source type is not polymorphic)
   95 |     dynamic_cast<B*>(ptr);
      | 
```

So we have to find a way to make the decision static_cast/dynamic_cast at compile time.

hageboeck commented on roofit/roofitcore/inc/RangeCast.h:
```diff
@@ -0,0 +1,133 @@
+   const_iterator end() const { return _inputRange.end(); }
+
+private:
+   Range_t const&_inputRange;
```
```suggestion
   Range_t const& _inputRange;
```

hageboeck reviewed (commented):
This looks quite good already, but consider removing the helper class.
```c++
if (templateParameter)
```
won't have any runtime overhead with any decent compiler, and both `dynamic_cast` and `static_cast` are syntactically correct code paths, so no `if constexpr` required.

guitargeek commented:
> This looks quite good already, but consider removing the helper class.
> 
> ```c++
> if (templateParameter)
> ```
> 
> won't have any runtime overhead with any decent compiler, and both `dynamic_cast` and `static_cast` are syntactically correct code paths, so no `if constexpr` required.

> This looks quite good already, but consider removing the helper class.
> 
> ```c++
> if (templateParameter)
> ```
> 
> won't have any runtime overhead with any decent compiler, and both `dynamic_cast` and `static_cast` are syntactically correct code paths, so no `if constexpr` required.

Thanks for the comments! Before I address the other comments: just one answer to this comment. No, it seems that `dynamic_cast<T*>` is not syntactically correct if T is not polymorphic, e.g. it has not virtual method.

You can for example see that this code doesn't compile:

```C++
struct A {
    void foo() {}
};

struct B : public A {
    void bar() {}
};

int main() {

    A* ptr = new B;
    dynamic_cast<B*>(ptr);
    delete ptr;
}
```
Here is the error:
```
collection_cast.cc: In function ‘int main()’:
collection_cast.cc:95:5: error: cannot ‘dynamic_cast’ ‘ptr’ (of type ‘struct A*’) to type ‘struct B*’ (source type is not polymorphic)
   95 |     dynamic_cast<B*>(ptr);
      | 
```

So we have to find a way to make the decision static_cast/dynamic_cast at compile time.

hageboeck commented:
> > This looks quite good already, but consider removing the helper class.
> > ```c++
> > if (templateParameter)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > won't have any runtime overhead with any decent compiler, and both `dynamic_cast` and `static_cast` are syntactically correct code paths, so no `if constexpr` required.
> 
> > This looks quite good already, but consider removing the helper class.
> > ```c++
> > if (templateParameter)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > won't have any runtime overhead with any decent compiler, and both `dynamic_cast` and `static_cast` are syntactically correct code paths, so no `if constexpr` required.
> 
> Thanks for the comments! Before I address the other comments: just one answer to this comment. No, it seems that `dynamic_cast<T*>` is not syntactically correct if T is not polymorphic, e.g. it has not virtual method.
> 
> You can for example see that this code doesn't compile:
> 
> ```c++
> struct A {
>     void foo() {}
> };
> 
> struct B : public A {
>     void bar() {}
> };
> 
> int main() {
> 
>     A* ptr = new B;
>     dynamic_cast<B*>(ptr);
>     delete ptr;
> }
> ```
> 
> Here is the error:
> 
> ```
> collection_cast.cc: In function ‘int main()’:
> collection_cast.cc:95:5: error: cannot ‘dynamic_cast’ ‘ptr’ (of type ‘struct A*’) to type ‘struct B*’ (source type is not polymorphic)
>    95 |     dynamic_cast<B*>(ptr);
>       | 
> ```
> 
> So we have to find a way to make the decision static_cast/dynamic_cast at compile time.

That's correct. I think whether or not you can use it depends on whether you want to use it only for RooFit collections or for other things as well?

eguiraud commented:
I just remembered that `TRangeDynCast` and `TRangeStaticCast` are already a thing in ROOT, meant for use with `TCollection`s. I looked them up and I think they are generic enough to be used with RooFit collections too (but they suffer from similar non-STL-compliance issues as the first version of this PR). Maybe generalize/improve those rather than adding a RooFit-specific version? So everybody benefits.

guitargeek commented:
Hi @eguiraud and @hageboeck, thanks a lot for your reviews!

I implemented almost everything you requested in the new fixup commits, except for two things:

1. https://github.com/root-project/root/pull/8769#issuecomment-891036930: I didn't generalize TRangeDynCast and TRangeStaticCast. They are tied to TCollection and fulfill a different purpose because of that. They are also part of `ROOT::Internal`, while my new functions are supposed to be also used by the RooFit users. I would suggest we leave them as they are, but if you think the new `static_range_cast` and `dynamic_range_cast` are useful for everyone we can move the new file if you want. But where?

2. https://github.com/root-project/root/pull/8769#discussion_r680892053: I didn't add the `assert(dynamic_cast)` to the `static_range_cast` because the performance penalty it introduces in debug mode is not probably not expected by non-RooFit developers.

3. https://github.com/root-project/root/pull/8769#issuecomment-890998817: I stuck with the detection of `dynamic_cast` vs. `static_cast` at compile time to also support non-polymorphic types.

Let me know where I still need to follow up, in particular concerning the more general location of the `RangeCast.h` file! Thanks.

hageboeck commented:
> Hi @eguiraud and @hageboeck, thanks a lot for your reviews!
> 
> I implemented almost everything you requested in the new fixup commits, except for two things:
> 
> 1. [[RF] New RangeCast file with dynamic_range_cast and static_range_cast #8769 (comment)](https://github.com/root-project/root/pull/8769#issuecomment-891036930): I didn't generalize TRangeDynCast and TRangeStaticCast. They are tied to TCollection and fulfill a different purpose because of that. They are also part of `ROOT::Internal`, while my new functions are supposed to be also used by the RooFit users. I would suggest we leave them as they are, but if you think the new `static_range_cast` and `dynamic_range_cast` are useful for everyone we can move the new file if you want. But where?

I was already expecting that it would be difficult to adapt those, but there will be ways to adapt them. I think it would be *better* if there's one good solution for all, but you have to decide how much time you want to invest.

> 2. [[RF] New RangeCast file with dynamic_range_cast and static_range_cast #8769 (comment)](https://github.com/root-project/root/pull/8769#discussion_r680892053): I didn't add the `assert(dynamic_cast)` to the `static_range_cast` because the performance penalty it introduces in debug mode is not probably not expected by non-RooFit developers.

There's not many loops out there in the world that will depend on this. If somebody uses it, they might be glad to have the extra safety. About the performance:
- When you run from the interpreter, you don't care about the performance of your loop as much.
- When you compile, you probably know what `NDEBUG` is.
- When you ACLIC-compile, I would have hoped that you get an automatic `NDEBUG` in release builds of ROOT, or when you `+O` a file. You don't, unfortunately. My preferred solution would have been to fix that and use the asserts, but it's your project now.

> 3. [[RF] New RangeCast file with dynamic_range_cast and static_range_cast #8769 (comment)](https://github.com/root-project/root/pull/8769#issuecomment-890998817): I stuck with the detection of `dynamic_cast` vs. `static_cast` at compile time to also support non-polymorphic types.
> 

Sounds good.

guitargeek commented:
Hi @hageboeck, thanks for the feedback!

Okay, I now took the time to hopefully get this range casting also work for the TCollections, so we don't have duplicate code.

Actually, there was an `assert` also in the old (TRangeStaticCast implementation)[https://github.com/root-project/root/blob/master/core/cont/inc/TCollection.h#L355]! So to stay consistent with TRangeStaticCast, I had to bring the `assert` in so that change request is also addressed now :) And we can also use the existing unit test in `testTypedIteration.cxx`.

eguiraud commented:
That's amazing @guitargeek !

Axel-Naumann commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+} // namespace ROOT
+
+template <typename T, typename WrappedIterator_t, bool isDynamic>
+class TypedIter {
```
All subsequent declarations should still be in the `ROOT` namespace.
Probably all by the `RangeCast` functions should even be in `ROOT::Internal`, or at least `ROOT::Detail`?

Axel-Naumann commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+#if (__cplusplus < 201700L)
+
+template <typename T, bool isDynamic = false>
+struct Cast {
```
Coding convection:
```suggestion
struct Cast_ {
```
or
```suggestion
struct RCast {
```

guitargeek replied:
Okay! I used `RCast` now.

Axel-Naumann commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+
+// For SFINAE-based checks for the existence of the `begin` and `end` methods.
+template <typename T>
+constexpr auto hasBeginEnd(int) -> decltype(std::declval<T>().begin(), std::declval<T>().end(), true)
```
```suggestion
constexpr auto hasBeginEnd(int) -> decltype(begin(std::declval<T>), end(std::declval<T>), true)
```
should allow C-style arrays, too:
```c++
TNamed n;
const TObject *arr[3] = {&n, &n, &n};
for (auto *named : static_range_cast<const TNamed*>(arr))
```

guitargeek replied:
Okay!

Axel-Naumann replied:
Would be great to have this as a test btw!

guitargeek replied:
Sure, I can make a test for that.

Axel-Naumann commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+   typename std::remove_reference<T>::type *operator->() { return &**this; }
+
+private:
+   WrappedIterator_t _iter;
```
Coding convention:
```suggestion
   WrappedIterator_t fIter;
```

guitargeek replied:
Okay!

Axel-Naumann commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+   template <typename R = Range_t, typename = std::enable_if_t<!std::is_const<R>::value>>
+   TypedIter<T, typename R::iterator, isDynamic> begin()
+   {
+      return _inputRange.begin();
```
```suggestion
      return begin(_inputRange);
```

guitargeek replied:
Okay!

Axel-Naumann commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+   template <typename R = Range_t, typename = std::enable_if_t<!std::is_const<R>::value>>
+   TypedIter<T, typename R::iterator, isDynamic> end()
+   {
+      return _inputRange.end();
```
```suggestion
      return end(_inputRange);
```

guitargeek replied:
Okay!

Axel-Naumann commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+   WrappedIterator_t _iter;
+};
+
+template <typename T, typename RangeReference_t, typename ConstIter_t, bool isDynamic>
```
I don't see that `ConstIter_t` is really needed as a dedicated template argument. You could deduce it as `decltype(cbegin(declval<RangeReference_t>()))` IIUC?

guitargeek replied:
Okay, that works too! I was using this `ConstIter_t`, because the `TCollection` doesn't implement `cbegin()` and also doesn't have a `const_iterator` typename. But all of these issues go away when using `std::cbegin()` and `std::cend()`, thanks I didn't think about that!

Axel-Naumann replied:
Right - the best way of doing this (allowing people to implement their own `cbegin` etc) is:
```
{
   using std::cbegin;
   cbegin(coll);
```
which makes `std::cbegin` available, but (thanks to argument dependent lookup) allows the code to find e.g. `ROOT::cbegin` for `ROOT::SomeCollection`

Axel-Naumann commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+};
+
+template <typename T, typename Range_t,
+          typename ConstIter_t = typename std::remove_reference<Range_t>::type::const_iterator>
```
Consider using the `decltype` notation I suggested above to enable this also on C-style arrays

guitargeek replied:
Okay!

Axel-Naumann commented on core/cont/inc/TCollection.h:
```diff
@@ -296,78 +298,11 @@ namespace Internal {
-};
-
-/// @brief TRangeStaticCast is an adaptater class that allows the typed iteration
+/// @brief TRangeStaticCast is an adapter function that allows the typed iteration
```
`TRangeStaticCast` used to be a class; it's awkward (be it for coding conventions alone) to change that to be a function. Do you think you can keep the class interface? Is there an advantage of making `RangeXYZCast` a function interface instead? I can see a downside, namely the limited temporary lifetime extension.

guitargeek replied:
Sure! I brought the class interface back for `TRangeStaticCast` and `TRangeDynCast`, but for the more general `static_range_cast` and `dynamic_range_cast` I could not do it (or now `StaticRangeCast` and `DynamicRangeCast` as you suggested to rename them).

The problem is that these functions have a second template parameter, which is the type of the input container. For classes, you'd need to pass the template argument explicitly which is what I don't want here. Or is there another solution for this?

guitargeek replied:
Or I can just put the functions only in RooFit, maybe that's less confusing because I will not introduce new free functions in `core`.

Axel-Naumann replied:
These functions are great and we should have them. But it'd be good if you can find a way to keep the `TCollection` versions classes - I suppose we know the iterator type after all?

Axel-Naumann commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+// Author: Jonas Rembser, CERN, Jul 2021
+
```
I'd propose to move this to `core/foundation/inc/ROOT` (IIUC it has no dependencies on anything ROOT?). It should be in `ROOT/` with a `.hxx` extension because it's namespaced and follows the new interface style. The filename should probably be `RRangeCast.hxx`

guitargeek replied:
Okay!

Axel-Naumann commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+};
+
+template <typename T, typename RangeReference_t, typename ConstIter_t, bool isDynamic>
+class RangeCast {
```
Coding conventions:
```suggestion
class RRangeCast {
```

guitargeek replied:
Okay!

Axel-Naumann reviewed (changes requested):
Great idea, thanks!

guitargeek commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+/// cast all objects in a RooAbsRange when iterating over them.
+/// Example:
+/// ~~~{.cpp}
+/// for(auto const* pdf : static_range_cast<RooAbsPdf*>(argSet)) {
```
Okay!

guitargeek commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+   RangeCast(RangeReference_t &&inputRange) : _inputRange{inputRange} {}
+
+   using const_iterator = TypedIter<T, ConstIter_t, isDynamic>;
+   const_iterator begin() const { return static_cast<Range_t const &>(_inputRange).begin(); }
+   const_iterator end() const { return static_cast<Range_t const &>(_inputRange).end(); }
```
Okay!

guitargeek commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+      if constexpr (isDynamic) {
+         return dynamic_cast<T>(*_iter);
+      } else {
+         assert(dynamic_cast<T>(*_iter));
```
Addressed! See comment above.

guitargeek commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+   const_iterator begin() const { return static_cast<Range_t const &>(_inputRange).begin(); }
+   const_iterator end() const { return static_cast<Range_t const &>(_inputRange).end(); }
+
+   // The non-const `begin` and `end` methods are disabled if the RangeCast wraps a const type.
```
Okay, actually it works fine with `std::begin` and `std::end`.

guitargeek commented on core/cont/inc/RangeCast.h:
```diff
@@ -0,0 +1,185 @@
+      }
+   }
+#endif
+   typename std::remove_reference<T>::type *operator->() { return &**this; }
```
The problem here is that `T` *can be* a reference: it is the type you cast to, so it is usually a reference or a pointer type.
Usually it is, so the code will work.

But yes, if you cast to a different value like `static_cast<double>` this is problematic because of the pointer to a temporary like you say. As far as I can tell, we will always have this problem with `operator->` no matter which one of the two is implemented in terms of the other.

Maybe we should just not implement `operator->`? It is not necessary for the use of `RRageCast` is range-based loops. I'll remove the `operator->` in my next commit, let me know if you prefer something else.

guitargeek commented:
Hi @Axel-Naumann, thanks a lot for all the comments! In my recent fixup commits, I tried to address all of them. Probably the only controversial change is that I removed the `static_range_cast` and `dynamic_range_cast` utility functions and put them into RooFit so I can keep the lower case name which I think is more standard for free functions.

guitargeek commented:
And it seems not all compilers are happy with the C-syle arrays, even though it worked fine on my machine with the newest GCC.

Axel-Naumann commented:
> I removed the static_range_cast and dynamic_range_cast utility functions and put them into RooFit so I can keep the lower case name which I think is more standard for free functions.

I'd still propose to keep `ROOT::RangeStaticCast` and `ROOT::RangeDynCast` in core/foundation - these functions are incredibly useful.

I'd also rename them as above, because not the range is static / dynamic but the cast.

Axel-Naumann commented:
> And it seems not all compilers are happy with the C-syle arrays, even though it worked fine on my machine with the newest GCC.

You will likely be able to work around that issue by introducing an additional constructor overload for RRangeCast, possibly taking a reference to an array,

guitargeek commented:
Hi @Axel-Naumann, thanks again for the comments!

I have now added the free `ROOT::RangeStaticCast` and `ROOT::RangeDynCast` functions, as well as an overload for each of them to deal with C-style arrays even with the old GCC 5.4, as used on the Ubuntu 16 test node.

guitargeek commented:
Now this PR is blocked by https://github.com/root-project/roottest/pull/762, which will fix the test failures.

guitargeek commented:
@phsft-bot build

guitargeek commented:
I think the failing test is unrelated to this PR.

hageboeck commented on core/foundation/inc/ROOT/RRangeCast.hxx:
```diff
@@ -0,0 +1,207 @@
+} // namespace Internal
+
+/// Wraps any collection that can be used in range-based loops and applies
+/// `static_cast<T>` or `dynamic_cast<T>` to each element.
```
```c++
/// \tparam T The type to ...
/// \tparam isDynamic Apply dynamic ...
...
```

guitargeek replied:
Okay!

Axel-Naumann commented:
Indeed: ROOT-performance-centos8-multicore/default really succeeded (Jenkins communication failure); 32bit is unrelated. Super nice, well done, congratulations @guitargeek !

hageboeck commented on core/foundation/inc/ROOT/RRangeCast.hxx:
```diff
@@ -0,0 +1,207 @@
+   iterator end() { return std::end(fInputRange); }
+
+private:
+   Range_t fInputRange;
```
Are you creating copies of containers here when called in the wrong way?

guitargeek replied:
Sure, if you call it in the wrong way and instantiate a `RRangeCast` with `Range_t` as a value type  explicitly, you can cause unnecessary moving (not copying, as the constructor has the signature `RRangeCast(Range_t &&inputRange)`).

However, if you use the `RangeStaticCast` or `RangeDynCast` functions that take a universal reference that is perfect-forwarded to `RRangeCast`, `Range_t` will always be a reference type.

hageboeck commented on core/foundation/inc/ROOT/RRangeCast.hxx:
```diff
@@ -0,0 +1,207 @@
+/// cast all objects in a RooAbsCollection when iterating over them.
+/// Example:
+/// ~~~{.cpp}
+/// for(auto *pdf : ROOT::RangeStaticCast<RooAbsPdf const*>(argSet)) {
```
Consider using a non-RooFit example for general ROOT usage.

guitargeek replied:
Okay!

Axel-Naumann commented on core/foundation/test/testRRangeCast.cxx:
```diff
@@ -0,0 +1,21 @@
+   const ClassA *arr[3] = {&b, &b, &b};
+   int n = 0;
+   for (auto *item : ROOT::RangeStaticCast<const ClassB *>(arr)) {
+      EXPECT_TRUE(item != nullptr);
```
Maybe even
```suggestion
      EXPECT_EQ(item, &b);
```

hageboeck replied:
👍 
and does something along the lines of the following make sense? The `auto` hides everything, so you might statically ensure that you get what you want.
```c++
static_assert(std::is_same<decltype(*item),ClassB>::value, ...);
```

Same for `dynamic`?

guitargeek replied:
Okay! I now consider also the type in `EXPECT_EQ` by doing:
```C++
EXPECT_EQ(item, static_cast<const ClassB *>(arr[n]));
```

Axel-Naumann commented on math/smatrix/test/testInversion.cxx:
```diff
@@ -14,9 +14,7 @@
-#ifndef N
-#define N 5
-#endif
+constexpr unsigned int N = 5;
```
Did I see this land in master independently?

guitargeek replied:
No, this landed in roottest. The `testInversion.cxx` code is duplicated in root and roottest, with only the code in roottest being actually used. That's maybe not the best situation, but I tried to keep the code syncronized for now. Maybe the code in root should be removed at some point.

hageboeck commented on core/cont/inc/TCollection.h:
````diff
@@ -296,78 +298,20 @@ namespace Internal {
-///    }
-/// ```
-template <class Containee> // Containee must derive from TObject.
-class TTypedIter : public TIter {
````
Is this breaking code that's somewhere out in the world? Or is this unlike, because it's internal?

guitargeek replied:
Unlikely, because it was not even used in ROOT, but yes I guess it's better to keep TTypedIter because it's not in the internal namespace.

hageboeck commented on core/cont/inc/TCollection.h:
```diff
@@ -296,78 +298,20 @@ namespace Internal {
-      return static_cast<Containee *>(obj);
-   }
-
+template <typename T, bool isDynamic>
```
Add doxygen.

guitargeek replied:
Ok!

hageboeck commented on core/cont/inc/TCollection.h:
```diff
@@ -296,78 +298,20 @@ namespace Internal {
 
-/// @brief TRangeStaticCast is an adaptater class that allows the typed iteration
+/// @brief TRangeStaticCast is an adapter class that allows the typed iteration
 /// through a TCollection.  This requires the collection to contains element
```
`contains --> contain`

Consider adding the
```c++
/// \tparam T ...
```

hageboeck replied:
Is it only for `TCollection`? I don't think so. Also, maybe update the code example below to not use the  asserts?

guitargeek replied:
True, the asserts are done in `TRangeStaticCast` itself! And indeed `TRangeStaticCast` is only for `TCollections`.

hageboeck commented on core/cont/inc/TCollection.h:
```diff
@@ -385,22 +329,13 @@ class TTypedIter : public TIter {
 } // namespace ROOT
 
-/// @brief TRangeDynCast is an adaptater class that allows the typed iteration
+/// @brief TRangeDynCast is an adapter class that allows the typed iteration
```
Same as above.

guitargeek replied:
Ok!

Axel-Naumann reviewed (approved):
Tiny comments otherwise LGTM!

guitargeek commented on roofit/roofitcore/inc/RooAbsCollection.h:
```diff
@@ -201,11 +201,27 @@ class RooAbsCollection : public TObject, public RooPrintable {
+    return _list.cend();
+  }
+
+  Storage_t::iterator begin() {
+    return _list.begin();
+  }
+
+  Storage_t::iterator end() {
+    return _list.end();
+  }
+
```
I will just remove the commit that introduces these new functions. They are not needed anymore, now that I use `std::begin()` and `std::cbegin()` in RRangeCast as Axel suggested.

guitargeek commented:
Hi @hageboeck, thanks for the comments! I have addressed them now. Is all the documentation I added here good enough so we can close https://github.com/root-project/root/issues/8421 after this PR is merged?

hageboeck commented on core/foundation/test/testRRangeCast.cxx:
```diff
@@ -0,0 +1,37 @@
+   const ClassA *arr[3] = {&b, &b, &b};
+   int n = 0;
+   for (auto *item : ROOT::RangeStaticCast<const ClassB *>(arr)) {
+      EXPECT_EQ(item, static_cast<const ClassB *>(arr[n]));
```
Here, you are checking the pointers. However, the types (i.e. that the casting happens) aren't checked. See this:
```c++
root [0] class A{} a; class B : public A {} b;
root [3] &a == &b
(bool) false
root [4] static_cast<A*>(&b) == &b
(bool) true
```
That's why I suggested the `static_assert`.

guitargeek replied:
Ah okay, I thought this was implicitly done with the pointer comparison, but I checked now that it doesn't cover it because you can compare `A*` with `B*`. Now I brought in the static assert.

hageboeck commented:
> Hi @hageboeck, thanks for the comments! I have addressed them now. Is all the documentation I added here good enough so we can close #8421 after this PR is merged?

I think it's getting there, but it's hard to say "yes". The real test is to run doxygen, and ask a member of the team to find the classes and how to use them.
